src/util/futurepool.rs

```diff
@@ -0,0 +1,227 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/// This mod implemented a wrapped future pool that supports `on_tick()` which is driven by
+/// tasks and is invoked no less than the specific interval.
+
+use std::fmt;
+use std::cell;
+use std::sync::{self, mpsc};
+use std::thread;
+use std::time;
+use futures::Future;
+use futures_cpupool as cpupool;
+
+use util;
+use util::time::Instant;
+use util::collections::HashMap;
+
+pub trait Context: fmt::Debug + Send + 'static {
+    fn on_tick(&mut self) {}
+}
+
+#[derive(Debug)]
+/// A wrapper to wrap context to provide `on_tick` feature.
+struct ContextOuter<T: Context> {
+    tick_interval: time::Duration,
+    /// TODO: Can be replace by UnsafeCell to eliminate the runtime cost of borrow checking.
+    inner: cell::RefCell<T>,
+    last_tick: cell::Cell<Option<Instant>>,
+}
+
+impl<T: Context> ContextOuter<T> {
+    fn new(context: T, tick_interval: time::Duration) -> ContextOuter<T> {
+        ContextOuter {
+            tick_interval,
+            inner: cell::RefCell::new(context),
+            last_tick: cell::Cell::new(None),
+        }
+    }
+
+    fn on_task_finish(&self) {
+        let now = Instant::now_coarse();
+        let last_tick = self.last_tick.get();
+        if last_tick.is_none() {
+            // set last_tick when the first future is resolved
+            self.last_tick.set(Some(now));
+            return;
+        }
+        if now.duration_since(last_tick.unwrap()) < self.tick_interval {
+            return;
+        }
+        self.last_tick.set(Some(now));
+        self.inner.borrow_mut().on_tick();
+    }
+}
+
+/// Each `ContextOuter` instance is invoked individually for each thread
+/// so that we mark it as Sync.
+unsafe impl<T: Context> Sync for ContextOuter<T> {}
```

Comment:
Is the unsafe key words necessary?

Reply:
Sync is an unsafe trait (since it is telling the compiler to trust the programmer, which is unsafe) so that when implementing it `unsafe` is required: https://doc.rust-lang.org/std/marker/trait.Sync.html